### ISSUE TITLE
Can't add data disks to azurerm_virtual_machine

### ISSUE THREAD

**Issue by nitdana** [hashicorp/terraform#6867]
Hi,

The issue here is about an error in creating and attaching data disks to azurerm_virtual_machine instances.
Any answer will be much appreciated!

**Version**: Terraform v0.6.16
**Affected Resource**: azurerm_virtual_machine.storage_data_disk
**Expected Behavior**: Create the data disk and attach it to the machine only after it exists.
**Actual Behavior**: A data disk is attached to the machine, but azure claims that it doesn't exist. I can see it in the virtual machine page, but when I search it in the relevant storage account -> storage container, it doesn't seem to appear.
It also causes the virtual machine to turn to "Failed" state and failes the terraform process.
**Important Factoids**: Azure Resource Manager
**Reproduce**:

1. Create the following resources (that a virtual machine in Azure Resource Manager depends on):

- azurerm_resource_group
- azurerm_virtual_network
- azurerm_subnet
- azurerm_storage_account
- azurerm_storage_container
- azurerm_network_interface

2. Create a virtual machine (azurerm_virtual_machine) with a data disk as shown in the relevant config
3. Run `terraform apply`

**Relevant config**:
```
variable "machine_name" {
    default = "MACHINE01"
}

resource "azurerm_network_interface" "Machine01NIC01" {
    name = "${var.machine_name}-if1"
    location = "${azurerm_resource_group.ResourceGroup01.location}"
    resource_group_name = "${azurerm_resource_group.ResourceGroup01.name}"

    ip_configuration {
        name = "testconfiguration1"
        subnet_id = "${azurerm_subnet.InternalSubnet01.id}"
        private_ip_address_allocation = "dynamic"
    }
}

resource "azurerm_virtual_machine" "Machine01" {
    name = "${var.machine_name}"
    location = "${azurerm_resource_group.ResourceGroup01.location}"
    resource_group_name = "${azurerm_resource_group.ResourceGroup01.name}"
    network_interface_ids = ["${azurerm_network_interface.Machine01NIC01.id}"]
    vm_size = "Standard_DS3"

    storage_image_reference {
        publisher = "Canonical"
        offer = "UbuntuServer"
        sku = "12.04.5-LTS"
        version = "latest"
    }

    storage_os_disk {
        name = "${var.machine_name}-os01"
        vhd_uri = "${azurerm_storage_account.StorageAccount01.primary_blob_endpoint}${azurerm_storage_container.StorageContainer01.name}/${var.machine_name}-os01.vhd"
        caching = "ReadWrite"
        create_option = "FromImage"
    }

    os_profile {
        computer_name = "hostname"
        admin_username = "admin"
        admin_password = "Password1234!"
    }

    os_profile_linux_config {
        disable_password_authentication = false
    }
    
    storage_data_disk {
        name = "${var.machine_name}-data01"
        vhd_uri = "${azurerm_storage_account.StorageAccount02.primary_blob_endpoint}${azurerm_storage_container.StorageContainer02.name}/${var.machine_name}-data01.vhd"
        create_option = "attach"
        disk_size_gb = "1023"
        lun = "0"
    }
    
    storage_data_disk {
        name = "${var.machine_name}-data02"
        vhd_uri = "${azurerm_storage_account.StorageAccount02.primary_blob_endpoint}${azurerm_storage_container.StorageContainer02.name}/${var.machine_name}-data02.vhd"
        create_option = "attach"
        disk_size_gb = "1023"
        lun = "1"
    }

}
```

Thanks!
Nitzan Dana

**Comment by stack72** [comment]
Hi @nitdana 

Apologies for the error here. Unfortunately you are running into the issue of using a restricted username :( When using the username `admin`, this will be the result:

```
{
    "error": {
        "code": "InvalidParameter",
        "target": "adminUsername",
        "message": "The Admin Username specified is not allowed."
    }
}
```

This is actually causing your failure. Please can you try another username and I can add some validation to the system

Paul

**Comment by nitdana** [comment]
@stack72 - Appreciate your answer!
I actually changed the username only for the example.
The username I'm using is not the problem here.

**Comment by stack72** [comment]
Very strange. Died the TF output show as failed?

**Comment by nitdana** [comment]
Yes, I get the following error:
```
Error applying plan:

1 error(s) occurred:

* azurerm_virtual_machine.Machine01: Error waiting for Virtual Machine (MACHINE01) to become available: unexpected state 'Failed', wanted target '[Succeeded]'

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

**Comment by nitdana** [comment]
@stack72  - Just got another error, when destroying an instance its os disk doesn't get deleted.
Should I open another issue about it?

**Comment by tasquith** [comment]
@nitdana FYI - the disk issue has already been raised as #6610

Cheers,

Tom

**Comment by nitdana** [comment]
@tasquith - Thanks!

**Comment by stack72** [comment]
Hi @nitdana 

So i just got back to look at this. From your examples:

```
storage_data_disk {
        name = "${var.machine_name}-data02"
        vhd_uri = "${azurerm_storage_account.StorageAccount02.primary_blob_endpoint}${azurerm_storage_container.StorageContainer02.name}/${var.machine_name}-data02.vhd"
        create_option = "attach"
        disk_size_gb = "1023"
        lun = "1"
    }
```

This means you are trying to Attach an existing blob to the Machine - is this correct? IF you want to create a new data_disk, then the `create_option = "Empty"`

That will work - i have tested this today and added an example to the readme

Paul

**Comment by nitdana** [comment]
Thanks @stack72 , 
I tried changing the attribute to "Empty" but the problem still occurs.
The new disk is attached, but nothing creates it and the machine is in "unavailable" state.

**Comment by timhaak** [comment]
I'm also busy testing this against the rc2 (Lots of fixes for azure :) ).

What seems to be happening is that the disk isn't created but the server still get's an entry pointing to a blob that doesn't exist. 

If you try and pre-create the disk it then says it's not allowed to resize while the server is live. I think if I could work out what size it expects the pre-created drive to be this will work. Unfortunately one is given in bytes and the other in gigs. 

I'm going to keep playing to see if I can work out what size to give the disk creation.

Let me know if there is anything I can test on my side. 

Though thanks for what's there already :)

**Comment by timhaak** [comment]
Ok seems to be 10^8 and not 1024^3.

For anyone else who would like a temporary work around. 

Just pre-create the disks. 

The config below is to create a 1023G disk 

resource "azurerm_storage_blob" "disk01" {
  name = "disk01.vhd"

  resource_group_name = "${azurerm_resource_group.production-ne.name}"
  storage_account_name = "${azurerm_storage_account.storage.name}"
  storage_container_name = "${azurerm_storage_container.prod.name}"

  type = "page"
  size = 1023000000000
}

**Comment by stack72** [comment]
Hi @nitdana (and @timhaak)

I am going to add an acceptance test to Terraform for this - I cannot recreate the issue you are seeing - so adding an acceptance test will help us make sure if it's a real issue

The example has the following VM config:

```
resource "azurerm_virtual_machine" "test" {
    name = "acctvm-%d"
    location = "West US"
    resource_group_name = "${azurerm_resource_group.test.name}"
    network_interface_ids = ["${azurerm_network_interface.test.id}"]
    vm_size = "Standard_A0"

    storage_image_reference {
	publisher = "Canonical"
	offer = "UbuntuServer"
	sku = "14.04.2-LTS"
	version = "latest"
    }

    storage_os_disk {
        name = "myosdisk1"
        vhd_uri = "${azurerm_storage_account.test.primary_blob_endpoint}${azurerm_storage_container.test.name}/myosdisk1.vhd"
        caching = "ReadWrite"
        create_option = "FromImage"
    }

    storage_data_disk {
        name          = "mydatadisk1"
        vhd_uri       = "${azurerm_storage_account.test.primary_blob_endpoint}${azurerm_storage_container.test.name}/mydatadisk1.vhd"
    	disk_size_gb  = "1023"
    	create_option = "empty"
    	lun           = 0
    }

    os_profile {
	computer_name = "hostname%d"
	admin_username = "testadmin"
	admin_password = "Password1234!"
    }

    os_profile_linux_config {
	disable_password_authentication = false
    }

    tags {
    	environment = "Production"
    	cost-center = "Ops"
    }
}
```

I should know within the next 10 minutes if this works as expected

Paul

**Comment by stack72** [comment]
Hi all

Just a FYI, I have added an acceptance test for this to show that we *can* create `storage_data_disks`

https://github.com/hashicorp/terraform/pull/7456

Paul

**Comment by timhaak** [comment]
@stack72 Thanks

I didn't use the creat_option="empty" going to test that now.

**Comment by timhaak** [comment]
Just for interest this worked for me with create_option = empty.

**Comment by stack72** [comment]
Hi @timhaak 

Just a FYI, you will need `Empty` for now until I fix the casing

P.
